### PR TITLE
From lazy avro

### PR DESCRIPTION
**Pull request by AlexeyRaga**
## Changes
An iteration on the experimental Lazy decoding:

- Add `FromLazyAvro` which is just like `FromAvro` except that it decodes directly from `LazyValue`  without converting to `Value`. In our cases with large and complicated messages, it delivers up to 25% of performance!

-  Generate `FromLazyAvro` instances when deriving data types from schemas.

- Add useful instances for  `EitherN` (`Functor`, `Bifunctor`, `Traversable`, `Bitraversable`,  `Foldable`, `Bifoldable`, `Applicative`, `Monad`)

### CODE REVIEW

File: src/Data/Avro/Decode/Lazy.hs [hunk 1]
```diff
@@ -23,50 +23,55 @@ module Data.Avro.Decode.Lazy
...
-import Data.Avro.Decode.Lazy.Deconflict as C
-import Data.Avro.FromAvro
+import           Data.Avro.Decode.Get
+import           Data.Avro.Decode.Lazy.Convert      (toStrictValue)
+import           Data.Avro.Decode.Lazy.Deconflict   as C
+import           Data.Avro.Decode.Lazy.FromLazyAvro
+import           Data.Avro.FromAvro
 
-import Debug.Trace
+import           Debug.Trace
```
Comment (Kheldar) [inline src/Data/Avro/Decode/Lazy.hs]: You probably don't want this.

File: src/Data/Avro/Decode/Lazy/FromLazyAvro.hs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+instance FromLazyAvro B.ByteString where
+  fromLazyAvro (T.Bytes b) = pure b
+  fromLazyAvro v           = badValue v "ByteString"
+
+instance FromLazyAvro BL.ByteString where
+  fromLazyAvro (T.Bytes b) = pure (BL.fromStrict b)
+  fromLazyAvro v           = badValue v "Lazy ByteString"
+
+instance FromLazyAvro Int where
+  fromLazyAvro (T.Int i) | (fromIntegral i :: Integer) < fromIntegral (maxBound :: Int)
```
Comment (Kheldar) [inline src/Data/Avro/Decode/Lazy/FromLazyAvro.hs]: What's with the cast to `Integer`?

File: src/Data/Avro/Deriving.hs [hunk 3]
```diff
@@ -185,7 +188,7 @@ mkStrictPrimitiveField _ field =
 -- Person { firstName :: Text }
 -- @
 -- You may want to enable 'DuplicateRecordFields' if you want to use this method.
-mkAsIsFieldName :: TypeName -> Field -> Text
+mkAsIsFieldName :: Text -> Field -> Text
```
Comment (Kheldar) [inline src/Data/Avro/Deriving.hs]: Why the change to a less specific type?

File: src/Data/Avro/Decode/Lazy/FromLazyAvro.hs (hunk 2 above)
Comment (AlexeyRaga) [reply src/Data/Avro/Decode/Lazy/FromLazyAvro.hs]: Same as `FromAvro` does, it is a copy. I haven't looked at it carefully (if at all), but I _presume_ that it may have something to do with the fact that `Int` is platform-specific.
_Maybe_ it is possible that the avro value that is encoded against the `Int` schema is actually bigger than the platform-specific `Int` at the decoding time?

But I don't know the answer.

File: src/Data/Avro/Deriving.hs (hunk 3 above)
Comment (AlexeyRaga) [reply src/Data/Avro/Deriving.hs]: For consistency. All the other functions take `Text`, and the `DerivingOptions` expects `Text -> Field -> Text` and not `TypeName -> Field -> Text`. Maybe this one was just forgotten at some point.

File: src/Data/Avro/Decode/Lazy/FromLazyAvro.hs (hunk 2 above)
Comment (Kheldar) [reply src/Data/Avro/Decode/Lazy/FromLazyAvro.hs]: Maybe consider using `Int32` and `Int64` specifically? `Integer` is very inefficient, so it'd be preferable to avoid going through it unnecessarily.

File: src/Data/Avro/Decode/Lazy.hs (hunk 1 above)
Comment (AlexeyRaga) [reply src/Data/Avro/Decode/Lazy.hs]: True, thanks.
